### DISCUSSION TITLE
`entry` action not triggering

### DISCUSSION THREAD

**Discussion by Perlkonig** [statelyai/xstate#1377] [Q&A] [answered]
I have an `entry` action that is not triggering when being jumped to from deeper in the hierarchy. Here's the minimal complete example:

```ts
import { Machine, interpret } from "xstate";

const machine = Machine(
    {
        initial: "myState",
        states: {
            myState: {
                id: "top",
                entry: [(context, event) => {console.log("ON ENTRY TRIGGERED");}],
                initial: "mySubState",
                states: {
                    mySubState: {
                        on: {
                            AGAIN: {
                                target: "#top",
                                internal: false
                            }
                        }
                    }
                }
            }
        }
    }
);
console.log("Starting service");
const s = interpret(machine);
s.start();
console.log("Issuing AGAIN event");
s.send("AGAIN")
```

When I run this file, I get the following output:

```txt
Starting service
ON ENTRY TRIGGERED
Issuing AGAIN event
```

Note that the entry event is not triggered the second time. Why not? As you can see, even adding `internal: false` doesn't change the behavior. Any guidance appreciated.

**Comment by davidkpiano** [comment, marked as answer]
The `internal: false` is on the `mySubState` substate, so it's saying that the _substate_ will be re-entered, not the parent state.

Move the `AGAIN` transition to the parent state and you'll see the difference.

**Reply by Perlkonig** [reply]
Perhaps I'm misunderstanding your suggestion. I can't put the transition on the parent state. In my actual code, this happens two levels deep, and only that state should receive an `AGAIN` event. Removing `internal: false` doesn't change anything (nor should it, since all transitions are apparently "external" by default). So clearly I'm not understanding how `entry` actions work. My understanding from the docs is that unless a transition is explicitly marked as `internal`, the `entry` actions should be triggered.

**Reply by davidkpiano** [reply]
When in doubt, move from `entry` actions transition `actions`. It might be a little repetitive, but you'll have the behavior you want.

**Reply by Perlkonig** [reply]
I apologize for being slow. I do appreciate your time. I'm having two struggles here: (1) just getting my code to work and (2) trying to understand why the code is working the way it is.

As for the first struggle, if I am understanding you correctly, you are suggesting I refactor this code as follows?

```ts
import { Machine, interpret } from "xstate";

const machine = Machine(
    {
        initial: "pseudoEntry",
        states: {
            pseudoEntry: {
                always: [
                    {
                        target: "myState",
                        actions: ["log"]
                    }
                ]
            },
            myState: {
                id: "top",
                initial: "mySubState",
                states: {
                    mySubState: {
                        on: {
                            AGAIN: {
                                target: "#top",
                                actions: ["log"]
                            }
                        }
                    }
                }
            }
        }
    },
    {
        actions: {
            log: (context, event) => {
                console.log("CODE TRIGGERED")
            }
        }
    }
);
console.log("Starting service");
const s = interpret(machine);
s.start();
console.log("Issuing AGAIN event");
s.send("AGAIN")
```

Giving me:

```txt
Starting service
CODE TRIGGERED
Issuing AGAIN event
CODE TRIGGERED
```

This does indeed work, but it is overly verbose and opaque. Which brings me to the second struggle. Why does `entry` not work as expected? Is this a bug with xstate? Or is this a bug with my understanding of `entry` actions? Does it also not work this way in SCXML?

According to the docs, "`entry` actions are executed upon entering a state." The doc sidebar labelled "When should I use transition vs. entry/exit actions?" says the following:

> An entry/exit actions means "execute this action **on any transition that enters/exits this state**". Use entry/exit actions when the action is only dependent on the state node that it's in, and not on previous/next state nodes or events.

This is precisely the semantics I'm going for. So I'd really prefer an `entry` event if at all possible.

**Reply by davidkpiano** [reply]
Can you rephrase the problem in real-world terms? As in, what is the real-world use-case for this? There might be a better way to model this.

**Reply by Perlkonig** [reply]
I'm happy to share the code, but I'm not sure how it will help. I accept that there are multiple ways of modelling things, and you've already shown me a way to get the behaviour I want. So regardless of the many different ways one *could* approach this problem, what I really want to understand is why it's not working as I expect. Given what the docs say, either (a) this is a bug, (b) the docs are not complete, or (c) the docs are yelling the answer at me and I am simply not seeing it. What is it about the example situation that means the `entry` action is not triggered the second time? And is this particular case explained in the docs somewhere?

For the record, the repository is [here](https://github.com/Perlkonig/utopia-backend). The machine definition is [here](https://github.com/Perlkonig/utopia-backend/blob/d3a287763dcd077a4124a72e0659b9c515a44e2a/src/index.ts#L523-L760). The state that is not working is `searching`, which is called again from `searching.idle.waiting`. There is most certainly a great deal of streamlining that could be done. I already am going back to use more transition actions where it works (but only strings seem to work in transition `actions`). But in this particular case, the semantics of the `setSearchStage` is perfect for an `entry` action. It doesn't matter where you're coming from, when you enter this state, it should do this action.

**Reply by Andarist** [reply]
As to **why** this works that way - we'd probably have to make an educated guess or track down SCXML mailing list and find a discussion about this.

As to if this is expected - yes. The best description of this behavior can be found [here](https://www.w3.org/TR/scxml/#N101D7) and also [here](https://www.w3.org/TR/scxml/#N103BF) as a description of `external` attribute:
> Determines whether the source state is exited in transitions whose target state is a descendant of the source state. 

So this only even matters if target states are **descendants** of the source state ("mySubState" in your example) and this is clearly not the case here. 

It conceptually makes sense though for your case - you are not exiting the region created by "myState", you are just moving within it, so you are not reentering it.

**Reply by davidkpiano** [reply]
> what I really want to understand is why it's not working as I expect

To visualize this:

<img width="686" alt="Screen Shot 2020-08-13 at 9 04 30 AM" src="https://user-images.githubusercontent.com/1093738/90137811-24045600-dd44-11ea-94d6-dd0e376b4479.png">

The two inner arrows are essentially the same. When you have `internal: false` in `mySubState`, that _only_ applies to `mySubState`. That is, there's no `completelyExitAllAncestors: true` attribute.

When you specify a target, SCXML is using that to determine which transitions are enabled and, from that, _resolve_ which state nodes should be active. Even if you specify `#top`, since that's a compound node, that will resolve to `#top` _and_ `.mySubState`, since that's the initial state of `#top`. So from the algorithm's perspective, it's as if you're self-transitioning.

**Comment by Perlkonig** [comment]
Thank you all for your time. Sorry for all the questions (and another one coming). You are correct that it's quite logical, but it's not explicitly said anywhere in the docs that I can find. I'll make a minor PR right quick. And it's too bad there's no `exitAllAncestors` feature. But I get it. Refactoring I go.

**Reply by Perlkonig** [reply]
PR #1380 submitted.

**Comment by singuerinc** [comment]
Hi, I'm confused here, why regular functions are called while assign is not?
I have a super simple example to demonstrate it, again, I complete missed why assign actions on the initial state are not called.

```js
import { createMachine, interpret, assign } from "@xstate/fsm";

const machine = createMachine({
  initial: "init",
  context: {
    count: 0
  },
  states: {
    init: {
      entry: [
        () => console.log("regular function called"),
        assign({
          count: () => {
            console.log("assing action never called, why?");
            return 1;
          }
        })
      ],
      on: {
        NEXT: "next"
      }
    },
    next: {
      entry: [
        assign({
          count: () => {
            console.log("assing action called, as expected.");
            return 2;
          }
        })
      ]
    }
  }
});

const service = interpret(machine).start();

service.send("NEXT");

// regular function called 
// assing action called, as expected.
```

**Reply by davidkpiano** [reply]
@singuerinc Can you please make an example in CodeSandbox? [Here's an XState FSM template](https://codesandbox.io/s/xstate-fsm-template-olx5f?file=/src/index.js).

**Reply by singuerinc** [reply]
There you go! https://codesandbox.io/s/lively-snowflake-8qty3

**Reply by davidkpiano** [reply]
@singuerinc That might be a bug. Can you create an issue? Thanks!

**Reply by singuerinc** [reply]
There you go, thanks! https://github.com/davidkpiano/xstate/issues/1659
